### PR TITLE
ci/mz_lsp_server: binary release process

### PR DESCRIPTION
**Pull request by joacoc**
### Motivation

This PR implements the CI process for the Language Server Protocol (LSP). The release process is almost identical to the [`mz`](https://github.com/MaterializeInc/materialize/tree/main/ci/deploy_mz) release process, with the following differences:

1. This process does not deploy a Docker image.
2. The Linux binary is built without using Docker ([this line](https://github.com/MaterializeInc/materialize/blob/8593e1c6731e042f24fd73d062743d9e53301695/ci/deploy_mz/linux.py#L32) is replaced by a `cargo build ..`).

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: ci/deploy_mz_lsp_server/deploy_util.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def upload_tarball(tarball: Path, platform: str, version: str) -> None:
+    s3_object = f"mz-lsp-server-{version}-{platform}.tar.gz"
+    boto3.client("s3").upload_file(
+        Filename=str(tarball),
+        Bucket=BINARIES_BUCKET,
+        Key=s3_object,
+    )
+    if "aarch64" in platform:
```
Comment (nrainer-materialize) [inline ci/deploy_mz_lsp_server/deploy_util.py:54]: Could you add a comment why we need this special handling? We need to upload the file with both names?

**Review by nrainer-materialize — COMMENTED** [review]
> The release process is almost identical to the [mz](https://github.com/MaterializeInc/materialize/tree/main/ci/deploy_mz) release process, with the following differences:

I am wondering if it is possible to reuse existing logic without duplicating the existing code. To give an example, could we generalize `deploy_util.py` by extracting the object name / prefix to a member variable (and wrapping the code in class) and use it both in mz release and mz_lsp_server? What do you think about it?

**Comment by joacoc** [comment]
@nrainer-materialize, yes it was my initial thought, but after a talk, I followed this approach as a starting point. It brings the benif of not messing up the `mz` process, but duplicates code. We could start by making `deploy_util.py` more generic. Do you know if this file can be shared between the two processes, `deploy_mz` and `deploy_mz_lsp_server`?

**Comment by nrainer-materialize** [comment]
> @nrainer-materialize, yes it was my initial thought, but after a talk, I followed this approach as a starting point. It brings the benif of not messing up the `mz` process, but duplicates code. We could start by making `deploy_util.py` more generic. Do you know if this file can be shared between the two processes, `deploy_mz` and `deploy_mz_lsp_server`?

Yeah, I am not suggesting to merge both workflows. My thought was whether it is possible to use files like `deploy_util.py` as util or generic files that can be used by both workflows. I think it should be possible to import this file in `deploy_mz_lsp_server` without problems. However, it might make sense to move it to a "neutral" location if it is supposed to serve both `deploy_mz` and `deploy_mz_lsp_server`.

I am convinced that we should avoid to duplicate code ("code clones") where possible without major downsides. Code duplication increases the amount of code to be maintained and increases the risk that bug fixes only get applied to some of the clones.

@joacoc: Though, I don't want to block your PR. Let me know if you disagree with this suggestion. And if you want to address this but need to get the PR merged rather soon, feel free to do it in a follow-up PR.

**Review by nrainer-materialize — APPROVED** [review]

**Comment by joacoc** [comment]
@nrainer-materialize makes total sense! I will create an issue to make `deploy_utils.py` more generic and follow up with another PR. Thank you! I forgot to also tag @benesch; He made the original one, and I may be missing something.

**Pull request merged**